### PR TITLE
Unify matmul pattern matching

### PR DESCRIPTION
**Pull request by Priya2698**
1. Adds alpha, beta parameters in matmul pattern matching
2. Unifies matching for matmul and matmul + bias.

Issue #1669.

### CODE REVIEW

**Comment by Priya2698** [comment]
!build

**Review by wujingyue — COMMENTED** [review]
LGTM! I'll defer approval to Jacob.

File: csrc/ir/utils.cpp [hunk 1]
```diff
@@ -1400,22 +1417,37 @@ bool matchMatmulBiasCast(
...
+  }
 
-  bias = bcast->input(0); // Bias tensor in fp32
   auto* bias_cast = dynamic_cast<UnaryOp*>(bias->definition());
 
   // The bias tensor and matmul inputs should be of the same dtype.
   NVF_ERROR(
       bias_cast == nullptr || bias_cast->getUnaryOpType() == UnaryOpType::Cast,
-      "Expected the bias tensor to be preceded by castOp before broadcasting.");
+      "Expected the bias tensor to casted to Float.");
```
Comment (jacobhinkle) [inline csrc/ir/utils.cpp:1450]: I might have already asked this, but does the `bias` tensor have to be same type as `A` and `B` for `at::addmm`?

File: csrc/ir/nodes.cpp [hunk 2]
```diff
@@ -425,18 +425,34 @@ std::vector<PolymorphicValue> UnaryOp::evaluate(
...
           ee.evaluate(mma_lhs, known_values).as<at::Tensor>().squeeze(-1);
       const auto b =
           ee.evaluate(mma_rhs, known_values).as<at::Tensor>().squeeze(0);
-      // Bias is of shape [M,].
-      const auto bias_tensor =
-          ee.evaluate(bias, known_values).as<at::Tensor>().unsqueeze(-1);
-      return {at::addmm(bias_tensor, a, b)};
+      auto bias_tensor = ee.evaluate(bias, known_values).as<at::Tensor>();
+
+      if (bias_tensor.dim() != a.dim()) {
```
Comment (jacobhinkle) [inline csrc/ir/nodes.cpp:445]: Technically, we could have a bias that is 3D and matmul output that is 2D and add them, relying on implicit broadcasting. So we should probably just explicitly check the case that `bias_tensor.dim() == a.dim()` or `bias_tensor.dim() == a.dim() - 1` else fail the pattern.

File: csrc/ir/utils.h [hunk 3]
```diff
@@ -66,7 +66,9 @@ bool matchMatmulBiasCast(
     const UnaryOp* cast_op,
     Val*& mma_lhs,
     Val*& mma_rhs,
-    Val*& bias);
+    Val*& bias,
+    Val*& alpha,
+    Val*& beta);
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/utils.h:71]: Yes, I have a separate PR for that but I can also have the changes here itself. Might simplify a few things.

File: csrc/ir/nodes.cpp [hunk 4]
```diff
@@ -425,18 +425,34 @@ std::vector<PolymorphicValue> UnaryOp::evaluate(
...
+        // Bias is of shape [M,].
+        bias_tensor = bias_tensor.unsqueeze(-1);
+      }
+
+      const auto alpha_tensor =
+          alpha ? toScalar(ee.evaluate(alpha, known_values)) : 1;
+      const auto beta_tensor =
+          beta ? toScalar(ee.evaluate(beta, known_values)) : 1;
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/nodes.cpp:453]: Ahh right, that's an error.

File: csrc/ir/utils.cpp [hunk 5]
```diff
@@ -1378,17 +1378,34 @@ bool matchMatmulCast(const UnaryOp* cast_op, Val*& mma_lhs, Val*& mma_rhs) {
...
+    Val*& bias,
+    Val*& alpha,
+    Val*& beta) {
   auto* binary = dynamic_cast<BinaryOp*>(cast_op->input(0)->definition());
   if (binary == nullptr || binary->getBinaryOpType() != BinaryOpType::Add) {
     return false;
   }
 
-  auto* mma = dynamic_cast<MmaOp*>(binary->input(0)->definition());
+  // Check for alpha in first input: alpha * (MmaOp(A, B))
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/utils.cpp:1398]: You're right. This pattern won't get matched.
Will make the change.

File: csrc/ir/utils.cpp (hunk 1 above)
Comment (Priya2698) [reply csrc/ir/utils.cpp:1450]: Yes, the dtypes need to be the same

File: csrc/ir/nodes.cpp (hunk 2 above)
Comment (Priya2698) [reply csrc/ir/nodes.cpp:445]: Only the second condition should occur, right? Matmul operands will be 3D and bias will be 2D.

File: csrc/ir/utils.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/ir/utils.cpp:1398]: This also goes for the first `BinaryOpType::Add`. biasEpilogue uses one format, but using `add` we could swap bias and mma.

File: csrc/ir/nodes.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/ir/nodes.cpp:445]: OK yeah I forgot about the broadcast. but we could then have a 4D or 5D "bias" being added to the matmul output. This current pattern would match it and unsqueeze to 5 or 6 dimensions, whereas it should just fail to match in those cases.

File: csrc/ir/utils.cpp [hunk 6]
```diff
@@ -1366,46 +1366,52 @@ void verifyMmaOpForEvaluation(
...
+// 6. alpha * A x B
+// Note: We assume the first operand to be the MmaOp output
+bool matchMatmulPatterns(
     const UnaryOp* cast_op,
     Val*& mma_lhs,
     Val*& mma_rhs,
     Val*& bias,
     Val*& alpha,
     Val*& beta) {
+  // Check if there may be a bias present.
```
Comment (jacobhinkle) [inline csrc/ir/utils.cpp:1384]: You might want to either initialize all the outputs to `nullptr` here, or (possibly better) return a struct containing these values. That way you can be sure that checking for `nullptr` really means that element was not matched.

Comment (jacobhinkle) [reply csrc/ir/utils.cpp:1384]: Another advantage of using a struct is that you don't have to update every call site if you extend the pattern in the future.

**Review by jacobhinkle — APPROVED** [review]
Looking good! Just one suggestion. Nice work.

File: csrc/ir/utils.cpp (hunk 6 above)
Comment (Priya2698) [reply csrc/ir/utils.cpp:1378]: Great point, added `MatmulInputs` to hold all the variables which are initialized to `nullptr`.

**Comment by Priya2698** [comment]
!build

File: csrc/ir/utils.h [hunk 7]
```diff
@@ -60,13 +60,20 @@ MmaOpDetails getMmaOpDetails(
 
 void verifyMmaOpForEvaluation(MmaOp* mma_op, DataType expected_input_dtype);
 
-bool matchMatmulCast(const UnaryOp* cast_op, Val*& mma_lhs, Val*& mma_rhs);
+struct MatmulInputs {
```
Comment (wujingyue) [inline csrc/ir/utils.h:63]: Is LinearLayerInputs a better name? MatmulInputs gives me the wrong impression that it's just a and b.

File: csrc/ir/utils.cpp [hunk 8]
```diff
@@ -1366,29 +1366,46 @@ void verifyMmaOpForEvaluation(
...
-  if (mma == nullptr) {
+// Possible combinations:
+// 1. A x B + C
+// 2. alpha * A x B + C
+// 3. A x B + beta * C
+// 4. alpha * A x B  + beta * C
+// 5. A x B
+// 6. alpha * A x B
+// Note: We assume the first operand to be the MmaOp output
+bool matchMatmulPatterns(const UnaryOp* cast_op, MatmulInputs* matmul_inp) {
```
Comment (wujingyue) [inline csrc/ir/utils.cpp:1377]: MatmulInputs&

**Pull request merged**